### PR TITLE
feat: build SML and keep it fresh

### PR DESCRIPTION
**Pull request by shuplenkov**
## Issue being fixed or feature implemented

In order to verify chain lock signatures and rotate validators, we need to build Simplified Masternode List (SML) and keep it fresh up to the latest chain locked height.

## What was done?

Obtain SML before starting ABCI server and update it on the latest chain lock update

## How Has This Been Tested?

With tests

## Breaking Changes

None

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -30,6 +30,7 @@ CORE_JSON_RPC_HOST=127.0.0.1
 CORE_JSON_RPC_PORT=9998
 CORE_JSON_RPC_USERNAME=dashrpc
 CORE_JSON_RPC_PASSWORD=password
+CORE_NETWORK=regtest
```
Comment (shumkov) [inline .env.example:33]: It's not just Core network, actually. It's the platform network as well. Let's call it just `NETWORK` like everywhere?

File: lib/core/LatestCoreChainLock.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  /**
+   *
+   * @param {ChainLock} [chainLock]
+   */
+  constructor(chainLock) {
+    super();
+
+    this.chainLock = chainLock;
+  }
+
```
Comment (shumkov) [inline lib/core/LatestCoreChainLock.js:13]: Let's add `getChainLock()`?

File: lib/core/SimplifiedMasternodeList.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
+const SimplifiedMNList = require('@dashevo/dashcore-lib/lib/deterministicmnlist/SimplifiedMNList');
```
Comment (shumkov) [inline lib/core/SimplifiedMasternodeList.js:1]: Do we really need to wrap it?

File: lib/core/SimplifiedMasternodeList.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+class SimplifiedMasternodeList {
+  constructor() {
+    this.simplifiedMNList = new SimplifiedMNList();
```
Comment (shumkov) [inline lib/core/SimplifiedMasternodeList.js:5]: We need `SimplifiedMNListStore`.

File: lib/core/waitForDMLSyncFactory.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    if (currentBlock < 1000) {
+      // wait for 1000 blocks height
+      do {
+        await wait(10000);
+
+        ({ result: currentBlock } = await coreRpcClient.getBlockCount());
+      } while (currentBlock < 1000);
+    }
+
+    // create SMLStore
```
Comment (shumkov) [inline lib/core/waitForDMLSyncFactory.js:41]: remove

File: lib/core/waitForDMLSyncFactory.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    let latestRequestedHeight = 1;
+    if (latestCoreChainLock.chainLock) {
+      const diff = await coreRpcClient.protx('diff', latestRequestedHeight, latestCoreChainLock.chainLock.height);
+      latestRequestedHeight = latestCoreChainLock.chainLock.height;
+
+      // add to SML
+      const simplifiedMNListDiff = new SimplifiedMNListDiff(diff, coreNetwork);
```
Comment (shumkov) [inline lib/core/waitForDMLSyncFactory.js:48]: Just network.

File: lib/core/waitForDMLSyncFactory.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        await wait(10000);
+
+        ({ result: currentBlock } = await coreRpcClient.getBlockCount());
+      } while (currentBlock < 1000);
+    }
+
+    // create SMLStore
+    let latestRequestedHeight = 1;
+    if (latestCoreChainLock.chainLock) {
+      const diff = await coreRpcClient.protx('diff', latestRequestedHeight, latestCoreChainLock.chainLock.height);
```
Comment (shumkov) [inline lib/core/waitForDMLSyncFactory.js:44]: rawDiff

File: .env.example [hunk 8]
```diff
@@ -31,6 +31,8 @@ CORE_JSON_RPC_PORT=9998
 CORE_JSON_RPC_USERNAME=dashrpc
 CORE_JSON_RPC_PASSWORD=password
 
+NETWORK=regtest
```
Comment (shumkov) [inline .env.example:34]: testnet

File: lib/core/waitForDMLSyncFactory.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+   *
+   * @typedef waitForDMLSyncFactory
+   *
+   * @returns {Promise<void>}
+   */
+  async function waitForDMLSync() {
+    let { result: currentBlock } = await coreRpcClient.getBlockCount();
+
+    if (currentBlock < 1000) {
+      // wait for 1000 blocks height
```
Comment (shumkov) [inline lib/core/waitForDMLSyncFactory.js:32]:
```suggestion
      // wait for 1000 blocks height to make sure that DML is enabled
```

File: lib/core/waitForDMLSyncFactory.js [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+      do {
+        await wait(10000);
+
+        ({ result: currentBlock } = await coreRpcClient.getBlockCount());
+      } while (currentBlock < 1000);
+    }
+
+    // create SMLStore
+    let simplifiedMNListStore;
```
Comment (shumkov) [inline lib/core/waitForDMLSyncFactory.js:41]: we need to pass it from the DI container

File: lib/core/LatestCoreChainLock.js [hunk 11]
```diff
@@ -16,8 +19,23 @@ class LatestCoreChainLock {
...
+  /**
+   *
+   * @return {ChainLock}
+   */
+  getChainLock() {
+    return this.chainLock;
+  }
 }
 
+LatestCoreChainLock.TOPICS = {
```
Comment (shumkov) [inline lib/core/LatestCoreChainLock.js:37]: EVENTS

File: lib/core/SimplifiedMasternodeList.js [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+      maxListsLimit: options.smlMaxListsLimit,
+    };
+  }
+
+  /**
+   * @param {SimplifiedMNListDiff[]} simplifiedMNListDiffArray
+   *
+   * @return SimplifiedMasternodeList
+   */
+  applyDiff(simplifiedMNListDiffArray) {
```
Comment (shumkov) [inline lib/core/SimplifiedMasternodeList.js:15]:
```suggestion
  applyDiffs(smlDiffs) {
```

File: lib/core/SimplifiedMasternodeList.js [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    return this;
+  }
+
+  /**
+   *
+   * @return {SimplifiedMNList|undefined}
+   */
+  getSimplifiedMNList() {
```
Comment (shumkov) [inline lib/core/SimplifiedMasternodeList.js:31]:
```suggestion
  getList() {
```

File: lib/core/errors/MissingChainlockError.js [hunk 14]
```diff
@@ -0,0 +1,11 @@
+class MissingChainlockError extends Error {
```
Comment (shumkov) [inline lib/core/errors/MissingChainlockError.js:1]:
```suggestion
class MissingChainLockError extends Error {
```

File: lib/core/errors/MissingChainlockError.js [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+  constructor() {
+    super('Chainlock was not retrieved from core');
```
Comment (shumkov) [inline lib/core/errors/MissingChainlockError.js:3]:
```suggestion
    super('ChainLock is required to obtain SML');
```

File: lib/core/waitForDMLSyncFactory.js [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+ * @param {RpcClient} coreRpcClient
+ * @param {SimplifiedMasternodeList} simplifiedMasternodeList
+ * @param {LatestCoreChainLock} latestCoreChainLock
+ * @param {number} smlMaxListsLimit
+ * @param {string} network
+ * @param {BaseLogger} logger
+ *
+ * @returns {waitForCoreSync}
+ */
+function waitForDMLSyncFactory(
```
Comment (shumkov) [inline lib/core/waitForDMLSyncFactory.js:20]: rename to waitForSMLSyncFactory

File: scripts/abci.js [hunk 17]
```diff
@@ -41,6 +41,12 @@ const errorHandler = require('../lib/errorHandler');
     await waitForCoreChainLockSyncFallback();
   }
 
+  const waitForDMLSync = container.resolve('waitForDMLSync');
+
+  logger.info('Obtaining DML...');
```
Comment (shumkov) [inline scripts/abci.js:46]:
```suggestion
  logger.info('Obtaining Simplified Masternode List...');
```

File: lib/core/waitForDMLSyncFactory.js [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+
+      simplifiedMNListDiffArray.push(new SimplifiedMNListDiff(rawDiff, network));
+    }
+
+    // add to SML
+    simplifiedMasternodeList.applyDiff(simplifiedMNListDiffArray);
+
+    latestRequestedHeight = chainLock.height;
+
+    logger.debug('Initial SML was set');
```
Comment (shumkov) [inline lib/core/waitForDMLSyncFactory.js:71]:
```suggestion
    logger.debug(`SML is initalized for height ${heightOffset} to ${chainLock.height}`);
```

File: lib/core/waitForDMLSyncFactory.js [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+
+        updatedSimplifiedMNListDiffArray.push(new SimplifiedMNListDiff(updatedRawDiff, network));
+      }
+
+      // add to SML
+      simplifiedMasternodeList.applyDiff(updatedSimplifiedMNListDiffArray);
+
+      latestRequestedHeight = updatedChainLock.height;
+
+      logger.debug('Updated SML');
```
Comment (shumkov) [inline lib/core/waitForDMLSyncFactory.js:96]:
```suggestion
      logger.debug(`SML is updated for heights ${latestRequestedHeight} to ${updatedChainLock.height}`);
```

**Review by 020401 — APPROVED** [review]

**Review by shumkov — APPROVED** [review]
LGTM! 👍

File: lib/core/SimplifiedMasternodeList.js [hunk 20]
```diff
@@ -26,10 +28,10 @@ class SimplifiedMasternodeList {
 
   /**
    *
-   * @return {SimplifiedMNList|undefined}
+   * @return {SimplifiedMNList}
```
Comment (shuplenkov) [inline lib/core/SimplifiedMasternodeList.js:31]: It can be undefined, right after the instance was created

Comment (shumkov) [reply lib/core/SimplifiedMasternodeList.js:31]: Yeah, but we usually do not define it as a type? It's just not present/defined.

**Review by shumkov — APPROVED** [review]

**Pull request merged**